run.c

```diff
@@ -222,7 +222,7 @@ void transformer(int token, int pos, Config* p, RunState* s, TransformerWeights*
     float* freq_cis_imag_row = w->freq_cis_imag + pos * head_size / 2;
 
     // forward all the layers
-    for(int l = 0; l < p->n_layers; l++) {
+    for(size_t l = 0; l < p->n_layers; l++) {
```

Comment:
Why is this necessary? p->n_layers is a small int, this introduces comparison of signed and unsigned, and for compiler optimizations it's better for loop counters to be signed.
```suggestion
    for(int l = 0; l < p->n_layers; l++) {
```

Reply:
It is the smallest change that fix the problem of overflow on places like `l*dim*dim`

Reply:
Right. Then I think it's better to replace

`int dim = p->dim;`

with 

`size_t dim = p->dim;`

instead

Reply:
No, it must be the first element, so the next ones will be automatically casted to size_t on multiplication

Reply:
I don't think so

> Otherwise, if the operand that has unsigned integer type has rank greater or equal to the rank of the type of the other operand, then the operand with signed integer type is converted to the type of the operand with unsigned integer type.

https://stackoverflow.com/questions/44055758/does-the-order-of-multiplication-variables-of-different-data-type-cause-differen

Reply:
Just checked on my compiler (gcc) and it behaves exactly as I thought:

```
#include <limits.h>
#include <stddef.h>
#include <stdio.h>

int main()
{
  int x = INT_MAX;
  int z = INT_MAX;
  size_t y = 4;

  printf("%lu\n", x*y);
  printf("%lu\n", y*x);
  printf("%d\n", x*z);
}
```

```
$ gcc int.c -o int && ./int
8589934588
8589934588
1
```

You can also play with format, gcc warns about argument type mismatch, printing nicely the type of the product